File: problems/345/345.reverse-vowels-of-a-string.py
#
# @lc app=leetcode id=345 lang=python3
#
# [345] Reverse Vowels of a String
#


# @lc code=start
class Solution:
    def reverseVowels(self, s: str) -> str:
        # leetcode -> eeoe -> eoee

        vowels = "aeiouAEIOU"
        res = list(s)
        n = len(s)

        def findLeft(i):
            while i < n:
                if s[i] in vowels:
                    break
                i += 1
            return i

        def findRight(i):
            while i >= 0:
                if s[i] in vowels:
                    break
                i -= 1
            return i

        l, r = findLeft(0), findRight(n - 1)

        while l < r:
            res[l], res[r] = res[r], res[l]
            l = findLeft(l + 1)
            r = findRight(r - 1)

        return "".join(res)


# @lc code=end

File: problems/345/345.reverse-vowels-of-a-string.java
/*
 * @lc app=leetcode id=345 lang=java
 *
 * [345] Reverse Vowels of a String
 */

// @lc code=start
class Solution {
    public String reverseVowels(String s) {
        // reverse order of vowels only
        // s = "leetcode" -> "eeoe" -> "eoee"
        // s = "leotcede"
        // 2 pointers swap
        // O(N) time, O(N) space

        int l = 0;
        int r = s.length() - 1;
        String vowels = "aeiouAEIOU";
        char[] ans = s.toCharArray();

        while (l < r) {
            while (l < r && vowels.indexOf(ans[l]) == -1) {
                l++;
            }

            while (l < r && vowels.indexOf(ans[r]) == -1) {
                r--;
            }

            // swap l and r
            char tmp = ans[l];
            ans[l] = ans[r];
            ans[r] = tmp;
            l++;
            r--;
        }

        return new String(ans);
    }
}
// @lc code=end
